mockld/event_data.go

```diff
@@ -0,0 +1,148 @@
+
+	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
+	"gopkg.in/launchdarkly/go-sdk-common.v2/ldvalue"
+)
+
+// Event is a JSON representation of an event. For convenience, this is stored as ldvalue.Value.
+type Event ldvalue.Value
+
+// Events is an array of events. This specialized type provides helper methods.
+type Events []Event
+
+// EventUser is a JSON representation of a user in an event. LaunchDarkly's JSON schema for users
+// in events is slightly different than users as SDK inputs.
+type EventUser ldvalue.Value
+
+func (e Event) Kind() string {
+	return ldvalue.Value(e).GetByKey("kind").StringValue()
+}
+
+// CanonicalizedJSONString transforms the JSON formatting of the event to make comparisons
+// reliable, as follows: 1. Drop the creationDate property, since its value is unpredictable.
+// 2. Call EventUser.CanonicalizedJSONString on user properties, if any. 3. Ensure that
+// all expected properties, if nullable, do appear in the object when null instead of being
+// omitted.
+func (e Event) CanonicalizedJSONString() string {
+	b := ldvalue.ObjectBuild()
+	for key, value := range ldvalue.Value(e).AsValueMap().AsMap() {
+		if key == "creationDate" {
+			continue // We won't try to make assertions about the timestamp, just omit it
+		}
+		if key == "user" {
+			b.Set(key, ldvalue.Raw([]byte(EventUser(value).CanonicalizedJSONString())))
+		} else {
+			b.Set(key, value)
+		}
+	}
+	var nullableProps []string
+	switch e.Kind() {
+	case "feature":
+		nullableProps = []string{"userKey", "user", "version", "variation", "reason"}
+	case "custom":
+		nullableProps = []string{"userKey", "user", "data", "metricValue"}
+	}
+	for _, k := range nullableProps {
+		if !ldvalue.Value(e).GetByKey(k).IsDefined() {
+			b.Set(k, ldvalue.Null())
+		}
+	}
+	return b.Build().JSONString()
+}
+
+func (e Event) AsValue() ldvalue.Value { return ldvalue.Value(e) }
+func (e Event) JSONString() string     { return helpers.CanonicalizedJSONString(e.AsValue()) }
+func (e Event) String() string         { return e.JSONString() }
+
+func (e *Event) UnmarshalJSON(data []byte) error {
+	var v ldvalue.Value
+	err := json.Unmarshal(data, &v)
+	*e = Event(v)
+	return err
+}
+
+func (es Events) JSONString() string {
+	data, _ := json.Marshal(es)
+	return string(data)
+}
+
+func SimpleEventUser(user lduser.User) EventUser {
+	// This function can be used when we're not testing anything related to private attributes
+	return ExpectedEventUserFromUser(user, servicedef.SDKConfigEventParams{})
+}
+
+func ExpectedEventUserFromUser(user lduser.User, eventsConfig servicedef.SDKConfigEventParams) EventUser {
+	// This simulates the expected behavior of SDK event processors with regard to redacting
+	// private attributes.
+	allJSON := helpers.AsJSONValue(user)
+	o := ldvalue.ObjectBuild()
+	var custom ldvalue.ObjectBuilder
+	var private []ldvalue.Value
+	allAttributes := append(allJSON.Keys(), allJSON.GetByKey("custom").Keys()...)
+	for _, attr := range allAttributes {
```

Comment:
I'm having some trouble understanding the intent of this loop. Would there be a way to split it into multiple passes, or otherwise refactor out some of the tests into named methods?

Reply:
Well, this is imitating code that we already implemented in https://github.com/launchdarkly/go-sdk-events which is the standard behavior of all of the LD SDKs regarding private attributes in events— as the comment says. And the behavior of private attributes is documented in every SDK, so I'd rather not repeat the explanation in detail here, and I don't want to make the logic extra long by adding unnecessary multiple passes. I can add a couple more comments though.

Reply:
Done - hope that's clearer.